### DISCUSSION TITLE
FIP-0: A proposal for making proposals

### DISCUSSION THREAD

**Discussion by varunsrin** [farcasterxyz/protocol#82] [FIP Stage 4: Finalized]
Title: A proposal for making proposals
Type: Process FIP 
Author: Varun Srinivasan (gh: @varunsrin , fc: @v) 

# Abstract

A proposal to make proposals that propose changes to the Farcaster ecosystem, and a process for approving or rejecting them.

![image](https://user-images.githubusercontent.com/528882/233211953-28c0a4b4-eeec-4b67-8bcd-0889fc6c6df9.png)

# Problem

Users and developers have no formal channel for proposing changes to the protocol. It unfolds somewhat haphazardly across Github Issues, Telegram groups and in Pull Requests. A standardized process would improve the pace and quality of protocol changes. 

# Specification

A FIP, or Farcaster Improvement Proposal, is a design document that proposes a change to Farcaster. They are inspired by [Ethereum's EIPs](https://eips.ethereum.org/EIPS/eip-1) and [Python's PEPs](https://peps.python.org/pep-0001/). There are three kinds of FIPS: 

1. **Process FIP** - adds or changes a process (e.g. FIP-4) 
2. **Standards FIP** - a standard for developers (e.g. FIP-1)
3. **Implementation FIP** - a change to the protocol implementation (e.g. FIP-2)

## Format

A finalized FIP must have the following: 

* **Title**: A short and memorable title under 50 characters.
* **Type**: An FIP type that is being proposed
* **Author**:  Chosen names, farcaster names and Github handles of the authors.
* **Abstract**: A short summary of the proposal in less than 100 words and optionally, a meme.
* **Problem**: An articulation of the problem you are trying to solve.
* **Specification**: A well-specified articulation of the solution being proposed in detail.
* **Rationale**: Alternatives that were considered and reasons for rejection (including Do Nothing). 
* **Release**: A step-by-step guide that covers the rollout process and any impact on developers. 

## Workflow 

A FIP moves through four discrete stages: Idea → Draft → Review → Finalized or Rejected 

#### 1. Idea 

Anyone can start an FIP by opening a discussion in [the ideas section](https://github.com/farcasterxyz/protocol/discussions/categories/fips-ideas-drafts) with the title `FIP: XYZ` and the label `fip-idea`. FIPs don't need to be in any format at this stage. Authors should focus on fleshing out the idea into a draft, building consensus with the community and getting buy-in from one or more [core contributors](https://github.com/orgs/farcasterxyz/teams/core). 

#### 2. Draft

Authors can move their FIPs into the draft stage by changing the label to `fip-draft`.  A Draft FIP is expected to follow the FIP format and address significant questions posed by the community.  An editor should be tagged in to move the draft into the final stage.

#### 3. Review

Editors can move Draft FIPs into review by moving the discussion into [the review section](https://github.com/farcasterxyz/protocol/discussions/categories/fips-in-review) with a title and number: `FIP-9999: Title` and adding the label `fip-review`. An FIP must have clear buy-in from the community and should be well-specified. If a code change is needed, an implementation should be provided. 

#### 4. Finalized 

Editors can finalize an FIP by adding the label `fip-finalized` after presenting it at the next developer call. If a code change is required, an implementation must be provided. 

#### 5. Cancelled

Editors or authors can cancel an FIP by adding the label `fip-cancelled` and closing the discussion at any time. 

## Editors 

Editors are volunteers who guide FIP's through the workflow and provide feedback. There is no formal process to becoming an editor. Anyone who has successfully finalized an FIP can volunteer to become an editor by asking the current editors.

The current editors are: 

- [Varun Srinivasan](https://github.com/varunsrin)

# Rationale

#### Why are FIPs lightweight and un-opinionated?

It is better to start simple and evolve as real-world usage guides design. What works for other communities may not be the right fit for us. 

#### Why don't you use X instead of Github discussions?

Github discussions was fast and easy to set up. If a better solution becomes apparent, we can migrate content later. 

# Release

This FIP should be finalized on April 26th and marked as FIP-0, pending feedback and approval from the community. In addition to this: 

1. Existing FIP's 1 through 4 should also be migrated to the format and stages specified in this document by April 26th
2. New FIP's can be proposed right away using this format.

**Comment by nichochar** [comment]
* I think that using semantic objects like github labels is better than modifying the title.

  So instead of `FIP: XYZ [Draft]`, use a label called `draft`

* Can you also define "core contributor" in this doc, or reference the definition?

* How does inactivity affect the status of a FIP?

**Reply by varunsrin** [reply]
1. agreed, updated 

2. defined by linking

3. no concrete proposal yet, at the discretion of the sponsoring core contributor

**Reply by nichochar** [reply]
LGTM

**Comment by varunsrin** [comment]
last call for feedback, this will be finalized in 48 hours

**Reply by davidfurlong** [reply]
lgtm!

**Comment by oreusa** [comment]
I saw a tweet that you are looking for an engineer in warpcast. My name is oreusa and I am Japanese. I have contacted you before from X, I emailed you with a proposal. The requirement is for a large number of Japanese shadowbans. My friend and I are both shadowbans. Here is a suggestion, How about building a control tower in Japan? We tell them to the terms and conditions to prevent spam. Because they cannot all read English and do not understand the terms and conditions. translation by creating a control tower there, We think that communicating in Japanese to the domestic market would be a good way to prevent spam. We don't want any reward. But the situation is so bad that people are shadowbanning without knowing the translation, I proposed this idea to improve the situation. My warpcast name is oreusa.eth I think warpcast is great, and judging by Frame's technology, I think it will expand in the future. I would appreciate your suggestions!
